railties/lib/rails/generators/rails/app/templates/.devcontainer/docker-compose.yml.tt

```diff
@@ -0,0 +1,78 @@
+services:
+  rails-app:
+    build:
+      context: ..
+      dockerfile: .devcontainer/Dockerfile
+
+    volumes:
+      - ../..:/workspaces:cached
+
+    # Overrides default command so things don't shut down after the process ends.
+    command: sleep infinity
+
+    networks:
+      - default
+
+    # Uncomment the next line to use a non-root user for all processes.
+    # user: vscode
+
+    # Use "forwardPorts" in **devcontainer.json** to forward an app port locally.
+    # (Adding the "ports" property to this file will not forward from a Codespace.)
+    ports:
+      - 45678:45678
+<%- if !devcontainer_dependencies.empty? -%>
+    depends_on:
+<%- devcontainer_dependencies.each do |dependency| -%>
+      - <%= dependency %>
+<%- end -%>
+<%- end -%>
+
+<%- if depends_on_system_test? -%>
+  selenium:
+    image: seleniarm/standalone-chromium
+    restart: unless-stopped
+    networks:
+      - default
+<%- end -%>
+
+<%- unless options.skip_active_job? && options.skip_action_cable? -%>
+  redis:
+    image: redis:7.2
+    restart: unless-stopped
+    networks:
+      - default
+    ports:
+      - 6379:6379
```

Comment:
Why do we expose Redis port to the host?

Reply:
See the comment above:

> # Use "forwardPorts" in **devcontainer.json** to forward an app port locally.
> # (Adding the "ports" property to this file will not forward from a Codespace.)

Reply:
> to forward an app port locally

Yeah, but why accessing internal (Docker) Redis locally?

I'm worried because that's a common situation when developers always have Redis running on their machine on the default port, so, trying to expose a Redis one on the same port would result in a conflict.

Reply:
Yes, I think you're right it is not needed. I am going to remove it.

Reply:
> See the comment above:
> 
> > ```shell
> > # Use "forwardPorts" in **devcontainer.json** to forward an app port locally.
> > # (Adding the "ports" property to this file will not forward from a Codespace.)
> > ```
> 
> But, in order to prevent third party access, I would suggest to bind all ports to `127.0.0.1`, e.g.,:
> 
> `"127.0.0.1:6379:6379"`

### Compose

Just in case, you are going to consider https://github.com/rails/rails/pull/50914#discussion_r1479818590.

I'm just realizing, when using Dev Containers, the ports are forward to host, even `forwardPorts` is not present. So, without binding to `"127.0.0.1:5432:5432"`, the database would be exposed to everyone in the network. 🥇 

We should avoid this!

### Dev Container CLI

Also, when there is no definition for port forwarding in compose file, e.g., `ports: 3000:3000` or better `"127.0.0.1:3000:3000"`, working with Dev Container CLI is not possible. The port will be not forward, even you define `forwardPorts` in order to access the Rails server. See https://github.com/devcontainers/cli/issues/186.

### Conclusion

Please, add port forwarding to `compose.yaml` file, not to `devcontainer.json`. 

By prepending it with `127.0.0.1:`, we can even secure expose database to host, in order to support tools like pgAdmin or other.